src/globus_sdk/tokenstorage/v2/validating_token_storage/validators.py

```diff
@@ -61,7 +62,7 @@ def after_retrieve(
         """
 
 
-class _OnlyBeforeValidator(TokenDataValidator):
+class _OnlyBeforeValidator(TokenDataValidator, ABC):
```

Comment:
I thought this wasn't needed.

`_OnlyBeforeValidator` already inherits from `abc.ABC`, and I thought it couldn't be instantiated unless a `before_retrieve()` is defined in a further subclass.


```suggestion
class _OnlyBeforeValidator(TokenDataValidator):
```

Reply:
I added these initially because PyCharm was complaining.

![image](https://github.com/user-attachments/assets/c7c790b3-076f-400c-99f8-94f51c0a9388)

---

You're correct that it's not needed. At runtime instantiating an `_OnlyBeforeValidator` fails because no `before_store` is defined whether `abc.ABC` is explicitly defined or not.

Conceptually though, this class flags that the subclasses are intended to be abstract as well, not concrete implementations. Given that it doesn't actually affect the class's mro (I just tested that locally) which is the only place that I could envision a negative runtime impact coming from, I'd argue this is just a clearer way to write it.

I don't feel passionately about this though; if folks disagree, ultimately I can change.

Reply:
Ah, then yes, please add that in. I, too, don't want to hear about it from PyCharm!

Reply:
It know it's probably irrational, but I don't like seeing `ABC` listed as an explicit base when it's already been inherited and the class is an ABC. It reads to me like a programming mistake, "Oh, you didn't realize but this was already an abstract class."

Consistent with that, I think that PyCharm lint is wrong. Cases like this one are not uncommon, in which an intermediate class can fill in some but not all of the details of a base class.

_That said_,
- I understand what the lint is catching (accidentally missing an abstract method)
- This has no runtime impact: `__mro__` is the same either way (oh, hey, you tested it too! :grin: )
- It's fine and this class is an internal implementation detail anyway

So it's okay, and I don't see a reason to make you fight with PyCharm, but I'll register the fact that visually I don't like it. :shrug:

Reply:
> I think that PyCharm lint is wrong. Cases like this one are not uncommon

It's also not uncommon to overlook an abstract method implementation, so I don't share the interpretation that the lint is wrong.